pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,438 @@
+			}
+
+			if !con.added {
+				con.added = true
+				s.addCon(con.ConID, con)
+				defer s.removeCon(con.ConID, con)
+			}
+		case pushEv := <-con.pushChannel:
+			// It is called when config changes.
+			// This is not optimized yet - we should detect what changed based on event and only
+			// push resources that need to be pushed.
+
+			// TODO: possible race condition: if a config change happens while the envoy
+			// was getting the initial config, between LDS and RDS, the push will miss the
+			// monitored 'routes'. Same for CDS/EDS interval.
+			// It is very tricky to handle due to the protocol - but the periodic push recovers
+			// from it.
+
+			err := s.pushConnection(con, pushEv)
+			if err != nil {
+				return nil
+			}
+
+		}
+	}
+}
+
+/*func mergeSubscribeUnsubsribe(resource []string, subscribe []string, unsubscribe []string) []string {
+	for i := 0; i < len(resource); i++ {
+		for j := 0; j < len(subscribe); j++ {
+			for k := 0; k < len(unsubscribe); k++ {
+
+			}
+		}
+	}
+}*/
+
+// Send with timeout
+func (conn *XdsConnection) sendDelta(res *xdsapi.DeltaDiscoveryResponse) error {
+	done := make(chan error)
+	t := time.NewTimer(SendTimeout)
+	go func() {
+		err := conn.deltaStream.Send(res)
+		done <- err
+		conn.mu.Lock()
+		if res.Nonce != "" {
+			conn.RouteNonceSent = res.Nonce
+		}
+		conn.mu.Unlock()
+	}()
+	select {
+	case <-t.C:
+		// TODO: wait for ACK
+		adsLog.Infof("Timeout writing %s", conn.ConID)
+		xdsResponseWriteTimeouts.Add(1)
+		return errors.New("timeout sending")
+	case err := <-done:
+		t.Stop()
+		return err
+	}
+}
+
+func subscribeToVHost(sidecarScope *model.SidecarScope, pushContext *model.PushContext,
+routeConfigName uint32, host string, namespace string) (*model.SidecarScope, error) {
+	if sidecarScope == nil {
+		return nil, fmt.Errorf("nil sidecar scope")
+	}
+
+	var sidecarConfig *networking.Sidecar
+	var ok bool
+
+	if sidecarScope.Config != nil {
+		sidecarCfg := sidecarScope.Config
+		sidecarConfig, ok = sidecarCfg.Spec.(*networking.Sidecar)
+		if !ok {
+			return nil, fmt.Errorf("invalid sidecar config")
+		}
+	} else {
+		sidecarConfig = &networking.Sidecar{
+			Egress: []*networking.IstioEgressListener{
+				{
+					//Bavery_QUESTION: This is likely wrong, but how do we figure out the port info? Specifically the protocol
+					Port: &networking.Port{
+						Number:   routeConfigName,
+						Protocol: "HTTP",
+					},
+				},
+			},
+		}
+
+	}
+
+	found := false
+
+	for _, egressPb := range sidecarConfig.Egress {
+		if egressPb.Port.Number != routeConfigName {
+			continue
+		}
+		found = true
+		egressPb.Hosts = append(egressPb.Hosts, host)
+		break
+	}
+	if !found {
+		sidecarConfig.Egress = append(sidecarConfig.Egress, &networking.IstioEgressListener{
+			Port: &networking.Port{
+				Number:   routeConfigName,
+				Protocol: "HTTP",
+			},
+		})
+	}
+
+	sdCfg := &model.Config{
+		Spec: sidecarConfig,
+	}
+	updatedSidecarScope := model.ConvertToSidecarScope(pushContext, sdCfg, namespace)
```

Comment:
Don't get it - just add a DynamicHosts list, with a lock - and a method to SidecarScope to change it.

Reply:
I can separate it into two lists, but:
* subscribing to resources -- this will require searching through both the sidecar CRD defined sidecar routes/hosts and the dynamic ones. 
* In the case of unsubscribing, what happens if a resource exists in the sidecar CRD defined resources? Do we modify the sidecar CRD resources in that case or are those set in stone? Note that last thing I knew, Envoy didn't have an implementation for unsubscribing anyway, so this might be a better question for later.
* If we keep them separate, I'm assuming BuildHTTPRoutes will be responsible for merging both lists (dynamic and sidecar CRD) every time an xDS response is sent?

Reply:
First problem is simple - doesn't seem hard to search in 2 sidecars.

Second point - we don't have to act immediately, we can wait until next 'add' or regular push. Since envoy doesn't even have an implementation...

On last point - simpler is better IMO. Let's solve it for hosts, make sure it works - and we solve most common use case. In 1.17 we can do it per route...

Reply:
Ok. Sounds good. Thanks. Was more concerned about any time overhead since this is all done mid-request.